ts/Stock/Indicators/MACD/MACDIndicator.ts

```diff
@@ -0,0 +1,545 @@
+
+
+        // Subtract each Y value from the EMA's and create the new dataset
+        // (MACD)
+        for (i = 1; i <= shortEMA.length; i++) {
+            if (
+                defined(longEMA[i - 1]) &&
+                defined(longEMA[i - 1][1]) &&
+                defined(shortEMA[i + (params.shortPeriod as any) + 1]) &&
+                defined(shortEMA[i + (params.shortPeriod as any) + 1][0])
+            ) {
+                MACD.push([
+                    shortEMA[i + (params.shortPeriod as any) + 1][0],
+                    0,
+                    null,
+                    shortEMA[i + (params.shortPeriod as any) + 1][1] -
+                        longEMA[i - 1][1]
+                ]);
+            }
+        }
+
+        // Set the Y and X data of the MACD. This is used in calculating the
+        // signal line.
+        for (i = 0; i < MACD.length; i++) {
+            xMACD.push(MACD[i][0]);
+            yMACD.push([0, null, MACD[i][3]]);
+        }
+
+        // Setting the signalline (Signal Line: X-day EMA of MACD line).
+        signalLine = (BaseSeries.seriesTypes.ema.prototype.getValues(
+            ({
+                xData: xMACD,
+                yData: yMACD
+            } as any),
+            {
+                period: params.signalPeriod,
+                index: 2
+            }
+        ) as any);
+
+        signalLine = (signalLine as any).values;
+
+        // Setting the MACD Histogram. In comparison to the loop with pure
+        // MACD this loop uses MACD x value not xData.
+        for (i = 0; i < MACD.length; i++) {
+            // detect the first point
+            if ((MACD[i] as any)[0] >= signalLine[0][0]) {
+
+                MACD[i][2] = signalLine[j][1];
+                yMACD[i] = [0, signalLine[j][1], MACD[i][3]];
+
+                if (MACD[i][3] === null) {
+                    MACD[i][1] = 0;
+                    yMACD[i][0] = 0;
+                } else {
+                    MACD[i][1] = correctFloat((MACD[i] as any)[3] -
+                    signalLine[j][1]);
+                    yMACD[i][0] = correctFloat((MACD[i] as any)[3] -
+                    signalLine[j][1]);
+                }
+
+                j++;
+            }
+        }
+
+        return {
+            values: MACD,
+            xData: xMACD,
+            yData: yMACD
+        } as IndicatorValuesObject<TLinkedSeries>;
+    }
+}
+
+interface MACDIndicator {
+    crispCol: typeof ColumnSeries.prototype['crispCol'];
+    currentLineZone: (string|null);
+    getColumnMetrics: typeof ColumnSeries.prototype['getColumnMetrics'];
+    graphmacd: (Highcharts.SVGElement|undefined);
+    graphsignal: (Highcharts.SVGElement|undefined);
+    macdZones: MACDZonesOptions;
+    nameComponents: Array<string>;
+    parallelArrays: Array<string>;
+    pointArrayMap: Array<string>;
+    pointClass: typeof MACDPoint;
+    pointValKey: string;
+    requiredIndicators: Array<string>;
+    signalZones: MACDZonesOptions;
```

Comment:
The following properties are not part of the extend call and should be move to the class pattern: currentLineZone, graphmacd, graphsignal, macdZones, signalZones.